clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -363,12 +370,14 @@ public ConsumerRebalanceListener listener() {
         private OffsetAndMetadata committed;  // last committed position
         private boolean paused;  // whether this partition has been paused by the user
         private OffsetResetStrategy resetStrategy;  // the strategy to use if the offset needs resetting
+        private boolean pendingUnsubscribe; // whether this partition is pending unsubscribe from its current user
```

Comment:
With this change, we may need to update some of the methods which read the `assignment` collection directly in `SubscriptionState` to filter out the methods. For example, don't we need to change `SubscriptionState.assignedPartitions()` and `SubscriptionState.pausedPartitions()`? Maybe we need some unit tests for `KafkaConsumer` to verify that unsubscribed partitions are not returned in `KafkaConsumer.assignment()`.